liquidity/liquidity.go

```diff
@@ -457,6 +499,160 @@ func (m *Manager) ForceAutoLoop(ctx context.Context) error {
+
+	// If we have already reached our total allowed number of in flight
+	// swaps we return early.
+	allowedSwaps := m.params.MaxAutoInFlight - summary.inFlightCount
+	if allowedSwaps <= 0 {
+		log.Debugf("%v autoloops allowed, %v in flight",
+			m.params.MaxAutoInFlight, summary.inFlightCount)
+
+		return nil
+	}
+
+	// Get all channels in order to calculate current hot wallet balance.
+	channels, err := m.cfg.Lnd.Client.ListChannels(ctx, false, false)
+	if err != nil {
+		return err
+	}
+
+	hotWalletBalance := btcutil.Amount(0)
+	for _, channel := range channels {
+		hotWalletBalance += channel.LocalBalance
+	}
+
+	// Since we're only autolooping-out we need to check if we are below
+	// the target, meaning that we already meet the requirements.
+	if hotWalletBalance <= m.params.EasyAutoloopTarget {
+		log.Debugf("hot wallet balance %v below target %v",
+			hotWalletBalance, m.params.EasyAutoloopTarget)
+		return nil
+	}
+
+	terms, err := m.cfg.LoopOutTerms(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Calculate the amount that we want to loop out. Check against the max
+	// swap amount of terms.
+	amount := hotWalletBalance - m.params.EasyAutoloopTarget
+	if amount > terms.MaxSwapAmount {
+		amount = btcutil.Amount(terms.MaxSwapAmount)
+	}
+
+	log.Debugf("easy autoloop: hot_wallet=%v, target=%v, "+
+		"attempting to loop out %v", hotWalletBalance,
+		m.params.EasyAutoloopTarget, amount)
+
+	// Sort the candidate channels based on descending local balance.
+	sort.Slice(channels, func(i, j int) bool {
+		max1 := math.Min(
+			channels[i].LocalBalance.ToUnit(btcutil.AmountSatoshi),
+			amount.ToUnit(btcutil.AmountSatoshi),
+		)
+		max2 := math.Min(
+			channels[j].LocalBalance.ToUnit(btcutil.AmountSatoshi),
+			amount.ToUnit(btcutil.AmountSatoshi),
+		)
+
+		return max1 > max2
+	})
+
+	// Start building that swap.
+	builder := newLoopOutBuilder(m.cfg)
+
+	channel := m.pickEasyAutoloopChannel(channels, terms, loopOut, loopIn)
+	log.Debugf("easy autoloop: picked channel %v with local balance %v",
+		channel.ChannelID, channel.LocalBalance)
+
+	swapAmt, err := btcutil.NewAmount(
+		math.Min(channel.LocalBalance.ToBTC(), amount.ToBTC()),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Override our current parameters in order to use the const percent
+	// limit of easy-autoloop.
+	easyParams := m.params
+	easyParams.FeeLimit = &FeePortion{
+		PartsPerMillion: defaultEasyAutoloopSwapFeePPM,
+	}
+
+	// Set the swap outgoing channel to the chosen channel.
+	outgoing := []lnwire.ShortChannelID{
+		lnwire.NewShortChanIDFromInt(channel.ChannelID),
+	}
+
+	suggestion, err := builder.buildSwap(
+		ctx, channel.PubKeyBytes, outgoing, swapAmt, true, easyParams,
+	)
+	if err != nil {
+		return err
+	}
+
+	swap := loop.OutRequest{}
+
+	switch t := suggestion.(type) {
+	case *loopOutSwapSuggestion:
+		swap = t.OutRequest
+	default:
+		return fmt.Errorf("unexpected swap suggestion type: %T", t)
+	}
+
+	// Dispatch a sticky loop out.
+	go m.dispatchStickyLoopOut(
```

Comment:
What if a previous call of `dispatchBestEasyAutoloopSwap` already dispatched a sticky loopout for a channel which backoffs? Wouldn't we happily do start a competing `dispatchStickyLoopOut`?

Reply:
a) If we hit our max inflight swap limit (default=`1`) we don't dispatch new swaps.

b) Check `pickEasyAutoloopChannel`: If there is an on-going swap on a channel we avoid picking that channel again for another easy-autoloop. Also if there was a failed swap on that channel we avoid picking it for some interval.

> Wouldn't we happily do start a competing `dispatchStickyLoopOut`?

So yes, we **could** happily start a new sticky loop out, but
- over a different outgoing channel
- over a channel that hasn't failed recently
- only if user has specified a `MaxAutoInFlight` > `1`